File: R/allDagsJonas.R
source("allDagsIntern.R")

allDagsJonas <- function(adj, row.names)
{
    # Input: adj. mat of a DAG with row.names, containing the undirected component that
    # should be extended
    # !!!! the function can probably be faster if we use partial orderings

    a <- adj[row.names, row.names]

    if(any((a + t(a)) == 1))
    {
        #warning("The matrix is not entirely undirected.")
        return(-1)
    }

    return(allDagsIntern(adj, a, row.names, NULL))
}


read_testcase <- function(filepath)
{
    lines <- readLines(filepath, warn = FALSE)

    adj <- list()
    row_names <- integer(0)
    mode <- NULL

    for(line in lines)
    {
        line <- trimws(line)

        if(line == "")
        {
            next
        }

        if(line == "adj:")
        {
            mode <- "adj"
            next
        }

        if(line == "row.names:")
        {
            mode <- "row_names"
            next
        }

        if(mode == "adj")
        {
            adj[[length(adj) + 1]] <- as.integer(strsplit(line, " +")[[1]])
        }
        else if(mode == "row_names")
        {
            row_names <- as.integer(strsplit(line, " +")[[1]])
        }
    }

    adj <- do.call(rbind, adj)

    return(list(
        adj = adj,
        row_names = row_names
    ))
}


script_dir <- getwd()
project_dir <- dirname(script_dir)

testcase_dir <- file.path(project_dir, "tests", "allDagsJonas", "Testcase")
output_dir <- file.path(project_dir, "tests", "allDagsJonas", "R_outputs")

if (!dir.exists(output_dir)) {
    dir.create(output_dir, recursive = TRUE)
}

output_file <- file.path(output_dir, "all_results.txt")
out <- file(output_file, open = "w")

for(i in 1:10000)
{
    filepath <- file.path(testcase_dir, paste0(i, ".txt"))
    values <- read_testcase(filepath)

    adj <- values$adj
    row_names <- values$row_names

    result <- allDagsJonas(adj, row_names)

    writeLines(paste0("Testcase ", i), out)

    writeLines("adj:", out)
    for(r in 1:nrow(adj)) {
        writeLines(paste(adj[r, ], collapse = " "), out)
    }

    writeLines("row.names:", out)
    writeLines(paste(row_names, collapse = " "), out)

    writeLines("Result:", out)
    if(length(result) == 1 && result[1] == -1)
    {
        writeLines("-1", out)
    }
    else if(is.null(result))
    {
        writeLines("empty", out)
    }
    else if(nrow(result) == 0)
    {
        writeLines("empty", out)
    }
    else
    {
        for(r in 1:nrow(result)) {
            writeLines(paste(result[r, ], collapse = " "), out)
        }
    }

    writeLines("", out)
}

close(out)

File: py_impl/allDagsJonas.py
import os
import numpy as np
from allDagsIntern import all_dags_intern


def all_dags_jonas(adj, row_names):
    adj = np.array(adj, dtype=int)
    row_names = np.array(row_names, dtype=int)

    a = adj[np.ix_(row_names - 1, row_names - 1)]

    if np.any((a + a.T) == 1):
        return -1

    tmp_init = np.empty((0, adj.size), dtype=int)
    return all_dags_intern(adj, a, row_names, tmp_init)


def read_testcase(filepath):
    adj = []
    row_names = []
    mode = None

    with open(filepath, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()

            if line == "":
                continue

            if line == "adj:":
                mode = "adj"
                continue

            if line == "row.names:":
                mode = "row_names"
                continue

            if mode == "adj":
                adj.append([int(x) for x in line.split()])
            elif mode == "row_names":
                row_names = [int(x) for x in line.split()]

    return np.array(adj, dtype=int), np.array(row_names, dtype=int)


#llm

script_dir = os.getcwd()
project_dir = os.path.dirname(script_dir)

testcase_dir = os.path.join(project_dir, "tests", "allDagsJonas", "Testcase")
output_dir = os.path.join(project_dir, "tests", "allDagsJonas", "Py_outputs")

if not os.path.exists(output_dir):
    os.mkdir(output_dir)

output_file = os.path.join(output_dir, "all_results.txt")

print("Reading testcases from:", testcase_dir)
print("Saving output to:", output_file)

with open(output_file, "w", encoding="utf-8") as f:
    for i in range(1, 10001):
        filepath = os.path.join(testcase_dir, f"{i}.txt")
        adj, row_names = read_testcase(filepath)

        result = all_dags_jonas(adj, row_names)

        f.write(f"Testcase {i}\n")

        f.write("adj:\n")
        for row in adj:
            f.write(" ".join(map(str, row)) + "\n")

        f.write("row.names:\n")
        f.write(" ".join(map(str, row_names)) + "\n")

        f.write("Result:\n")
        if isinstance(result, int) and result == -1:
            f.write("-1\n")
        elif result.shape[0] == 0:
            f.write("empty\n")
        else:
            for row in result:
                f.write(" ".join(map(str, row)) + "\n")

        f.write("\n")

print("Done. Output saved to:", output_file)